src/System.Runtime/src/System/HashCode.cs

```diff
@@ -0,0 +1,406 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Runtime.CompilerServices;
+
+namespace System
+{
+    public struct HashCode
+    {
+#   if SYSTEM_HASHCODE_TESTVECTORS
+        private static readonly uint _seed = 0;
+#   else
+        private static readonly uint _seed = unchecked((uint)new Random().Next(int.MinValue, int.MaxValue));
+#   endif
+
+        private const uint Prime1 = 2654435761U;
+        private const uint Prime2 = 2246822519U;
+        private const uint Prime3 = 3266489917U;
+        private const uint Prime4 = 0668265263U;
+        private const uint Prime5 = 0374761393U;
+
+        private uint _v1, _v2, _v3, _v4;
+        private uint _queue1, _queue2, _queue3;
+        private uint _length;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int Combine<T1>(T1 value1)
+        {
+            unchecked
+            {
+                var hc1 = (uint)(value1?.GetHashCode() ?? 0);
```

Comment:
Oddly I had this discussion with my brother today. He basically argued what you said. However,

* If your application differentiates between null and zero/unit, using zero in this case will result in a collision (which will swiftly fail an equals check). This results in marginally decreased performance.
* If your application does not differentiate (e.g. empty list = null) this will result in the surprising behavior of the same value being bucketed separately. This results in surprising behavior.

I don't like surprising other developers, so I'm quite strongly in the `0` camp.

Reply:
EqualityComparer.Default agrees with you. zero it is.